api/v1beta2/ratelimitpolicy_types.go

```diff
@@ -191,12 +200,32 @@ func (s *RateLimitPolicyStatus) Equals(other *RateLimitPolicyStatus, logger logr
 	return true
 }
 
+func (s *RateLimitPolicyStatus) GetObservedGeneration() int64 {
+	return s.ObservedGeneration
+}
+
+func (s *RateLimitPolicyStatus) SetObservedGeneration(o int64) {
+	s.ObservedGeneration = o
+}
+
 func (s *RateLimitPolicyStatus) GetConditions() []metav1.Condition {
 	return s.Conditions
 }
 
 var _ kuadrant.Policy = &RateLimitPolicy{}
 var _ kuadrant.Referrer = &RateLimitPolicy{}
+var _ reconcilers.ObjectWithStatus = &RateLimitPolicy{}
+
+func (r *RateLimitPolicy) GetStatusObject() reconcilers.StatusObject {
+	return &r.Status
+}
+
+func (r *RateLimitPolicy) SetStatusObject(statusObject reconcilers.StatusObject) {
+	switch status := statusObject.(type) {
+	case *RateLimitPolicyStatus:
+		r.Status = *status
```

Comment:
I am not sure about this silently failing if the wrong kind of `StatusObject` is passed in, it seems wrong. At the same time I don't think just logging an error message is enough.

Reply:
Maybe we could return `error` if as a `default` case (nothing matched) and we set the status to `nil` ? we could bubble up that error and decide given the setter expectations what to do

Reply:
before considering returning an `error`, any proposal?

Reply:
My mind is starting to go to extremes and I really don't know how well it suits. I am thinking that raising a panic might be the correct approach. If we return an error the reconcile ends up doing possibly twos things. One logs the error message, and two, reschedule the reconcile to try again. But if the wrong type is being past in, it was most likely an error caused by a developer in the code base where no matter how many times it is reschedule, reconcile, it will never pass. 

I know causing the panic is extreme and not really done in operators, but I don't see how it would ever natural recover. 

I do wonder what this would look like if the function is made more generalized  and taken off the struct its self. But possible would have other issue, along with some of the same issue that I can currently visualize in my head.

Reply:
thanks for the ideas... let me think and try something else. I reckon the type assertion is a bad smell.